server/data.py

```diff
@@ -7,19 +7,47 @@
 import json
 import utils
 import re
+import google.auth.transport.requests
+import google.auth
+import time
 
 class Datafetch(object):
 
     def _init_tabix(self):
         self.vcfs_imputed = self.conf['vcf_files']['imputed_data']
         self.vcfs_chip = self.conf['vcf_files']['chip_data']
-
+    
     def _init_db(self):
         self.conn = defaultdict(lambda: sqlite3.connect(self.conf['sqlite_db']))
 
     def _init_geo_data(self):
         with open(self.conf['geojson'], 'r') as f:
             self.geo_data = geojson.load(f)
+    
+    def _init_gcs_auth(self):
+        self.read_vcf_from_bucket = self.conf['vcf_files']['read_from_bucket']
+        if self.read_vcf_from_bucket:
+            self.token_retrieval_time = 0
+            self._refresh_gcs_auth_token()
+    
+    def _update_token_retrieval_time(self):
+        self.token_retrieval_time = time.time()
+    
+    def _refresh_gcs_auth_token(self):
```

Comment:
Great solution! My comments here are details but just thinking if it would make sense to call this function something like _refresh_gcs_auth_token_if_needed and then the "if self.read_vcf_from_bucket" check would be done in this function instead of checking that first before calling this function. Also to streamline, I think the _init_gcs_auth function is not needed and this kind of _refresh_gcs_auth_token_if_needed function is enough to do the whole job?

Reply:
yeah totally agree, this would be more clear indeed!